osmnx/simplify.py

```diff
@@ -354,4 +384,165 @@ def clean_intersections(G, tolerance=15, dead_ends=False):
     # get the centroids of the merged intersection polygons
     unified_intersections = gpd.GeoSeries(list(buffered_nodes))
     intersection_centroids = unified_intersections.centroid
-    return intersection_centroids
+
+    # To make things simpler, every edge should have a geometry
+    # (this avoids KeyError later when choosing between several
+    #  geometries or when the centroid inherits a geometry directly)
+    for u, v, data in G.edges(keys=False, data=True):
+        if 'geometry' not in data:
+            # if it doesn't have a geometry attribute, the edge is a straight
+            # line from node to node
+            x1 = G.nodes[u]['x']
+            y1 = G.nodes[u]['y']
+            x2 = G.nodes[v]['x']
+            y2 = G.nodes[v]['y']
+            data['geometry'] = LineString([(x1, y1), (x2, y2)])
+
+    # Let's first create the Graph first
+    G__ = nx.DiGraph(name = G.graph['name'],
```

Comment:
Why did you choose nx.DiGraph here? OSMnx typically works with nx.MultiDiGraph as parallel edges sometimes occur in street networks. Ideally, this would consistently use nx.MultiDiGraph unless there's a clear constraint.

Reply:
I think I chose to use a `nx.DiGraph` here, because it's simpler to add the edges, as I don't have to set the key parameter. I do cast it to a nx.MultiDiGraph on the return statement `return nx.MultiDiGraph(G__)`, but yes it might be better to use `nx.MultiDiGraph` from the start to make it consistent. I'll make that update.

Reply:
Thanks. We'd want to make sure we capture any parallel edges from the get-go (including their `key` attributes), or they could get lost in translation.

Reply:
I think these will be always lost because effectively we're also clustering (or merging) the edges, not just the nodes.

Reply:
Yes, I wasn't sure about this at first, and in fact, reviewing the code now, I'm only scanning the first layer (k = 0) of the input nx.MultiDiGraph, when merging the attributes of multiple edges between two centroids. Trying to recall why I coded it this way:

```python
    edges_attrs = [ G.edges[edge[0], edge[1], 0] for edge in out_edges_neighbor ]
```

Reply:
I sketched out a quick and dirty example... edges A and B are parallel edges (in the graph theory definition) and thus necessitate a MultiDiGraph rather than a DiGraph to model them. We'd want to retain both A and B in the cleaned graph, but if we use a DiGraph at some point in the workflow we'll lose one or the other.

![Untitled-1](https://user-images.githubusercontent.com/4977197/59113357-db1c6a80-8912-11e9-9d07-d7504ccc938a.png)

Reply:
So far, given two centroids (large nodes) and the corresponding members (small nodes), the current behavior of `clean_intersections` is to create a **single** new edge in the new graph, connecting the two large circles. And the attributes of this edge have to be chosen in some fashion among the multiple edges that exist between the members of each cluster.

But a better idea, might be to add as many new edges between the two new nodes (centroids) as there are edges between the members of each cluster in the original graph. I think with this solution we would most of the edge information in the original graph.

![osmnx_clean_intersections](https://user-images.githubusercontent.com/11331882/59114075-47609380-893e-11e9-996c-b965b843b546.png)

Reply:
> But a better idea might be to add as many new parallel edges between the two centroids as there are "old" edges.

I'm not sure about that. I think adding a single new edge is the right approach. However, we don't want to lose parallel edges from *non-clustered* nodes, such as in the example image I gave. Can we continue using the current single-edge implementation you've developed, but do a quick check for parallel edges in the original graph, and add them if found?

Reply:
Yeah, I can check for parallel edges between clusters with a single member and try to keep those. I need to find a test case, but it shouldn't be to hard to make these changes. I'll let you know.

Reply:
@gboeing I've implemented these changes. I've added/updated existing comments and also cleaned up the code a little bit (this includes the addition of an inner function 'fix_geometry' -- which is a bit ugly, but avoids repetition and makes the main loop easier to read). Please let me know what you think.